README.txt
This is a little extension for typo3 which provides 4 views for Events.

Events can be shown in a calendar(month) view, category view, list view with images or in boxes.




Icons by:
#########


------------------
iCandies Icon Set (60 icons)
Designed by IconEden (http://www.iconeden.com) and released especially for Smashing Magazine and its readers.
------------------

Dear friends,

thank you for downloading this file.

This freebie has been brought to you by SmashingMagazine.com.
You can freely use it for both your private and commercial projects, including software, online services, templates and themes.
The set may not be resold, sublicensed or rented. The set may not be offered for free downloading from websites other than SmashingMagazine.com.
Please link to the article in which this freebie was released if you want to spread the word.

Smashing Magazine Team,
www.smashingmagazine.com


ChangeLog
2011-05-24 Florian Schuhmann  <schuhmann.florian@googlemail.com>

	* initial code generated with kickstarter


ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "tx_bildungsangebote".
#
# Auto generated 24-05-2011 10:32
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Bildungsangebote',
	'description' => '',
	'category' => 'plugin',
	'author' => 'Florian Schuhmann',
	'author_email' => 'schuhmann.florian@googlemail.com',
	'shy' => '',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'module' => '',
	'state' => 'stable',
	'internal' => '',
	'uploadfolder' => 1,
	'createDirs' => '',
	'modify_tables' => '',
	'clearCacheOnLoad' => 1,
	'lockType' => '',
	'author_company' => '',
	'version' => '0.0.0',
	'constraints' => array(
		'depends' => array(
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:27:{s:9:"ChangeLog";s:4:"17b7";s:10:"README.txt";s:4:"ee2d";s:12:"ext_icon.png";s:4:"7516";s:17:"ext_localconf.php";s:4:"01f6";s:14:"ext_tables.php";s:4:"1a76";s:14:"ext_tables.sql";s:4:"2328";s:15:"flexform_ds.xml";s:4:"6aef";s:38:"icon_tx_txbildungsangebote_angebot.png";s:4:"d011";s:34:"icon_tx_txbildungsangebote_cat.png";s:4:"2653";s:13:"locallang.xml";s:4:"83aa";s:16:"locallang_db.xml";s:4:"ca2e";s:7:"tca.php";s:4:"66be";s:19:"doc/wizard_form.dat";s:4:"3e63";s:20:"doc/wizard_form.html";s:4:"7c7d";s:14:"pi1/ce_wiz.png";s:4:"706e";s:39:"pi1/class.tx_txbildungsangebote_pi1.php";s:4:"c05c";s:59:"pi1/class.tx_txbildungsangebote_pi1_addFieldsToFlexForm.php";s:4:"a559";s:47:"pi1/class.tx_txbildungsangebote_pi1_wizicon.php";s:4:"c697";s:13:"pi1/clear.gif";s:4:"cc11";s:35:"static/css/SpryCollapsiblePanel.css";s:4:"b527";s:25:"static/css/stylesheet.css";s:4:"0aad";s:30:"static/images/akkordion_bg.png";s:4:"d047";s:27:"static/images/bg_colums.png";s:4:"71e5";s:20:"static/images/h2.png";s:4:"05d9";s:20:"static/images/li.gif";s:4:"7e84";s:33:"static/js/SpryCollapsiblePanel.js";s:4:"575a";s:29:"static/js/SpryTabbedPanels.js";s:4:"34ed";}',
);

?>

ext_tables.php
<?php
if (!defined('TYPO3_MODE')) {
	die ('Access denied.');
}

t3lib_extMgm::allowTableOnStandardPages('tx_txbildungsangebote_cat');

$TCA['tx_txbildungsangebote_cat'] = array (
	'ctrl' => array (
		'title'     => 'LLL:EXT:tx_bildungsangebote/locallang_db.xml:tx_txbildungsangebote_cat',		
		'label'     => 'title',	
		'tstamp'    => 'tstamp',
		'crdate'    => 'crdate',
		'cruser_id' => 'cruser_id',
      	'sortby' => 'sorting',  
		'default_sortby' => 'ORDER BY crdate',	
		'delete' => 'deleted',	
		'enablecolumns' => array (		
			'disabled' => 'hidden',
		),
		'dynamicConfigFile' => t3lib_extMgm::extPath($_EXTKEY).'tca.php',
		'iconfile'          => t3lib_extMgm::extRelPath($_EXTKEY).'icon_tx_txbildungsangebote_cat.png',
	),
);


t3lib_extMgm::allowTableOnStandardPages('tx_txbildungsangebote_angebot');


t3lib_extMgm::addToInsertRecords('tx_txbildungsangebote_angebot');

$TCA['tx_txbildungsangebote_angebot'] = array (
	'ctrl' => array (
		'title'     => 'LLL:EXT:tx_bildungsangebote/locallang_db.xml:tx_txbildungsangebote_angebot',		
		'label'     => 'title',	
		'tstamp'    => 'tstamp',
		'crdate'    => 'crdate',
		'cruser_id' => 'cruser_id',
		'sortby' => 'sorting',  
		'delete' => 'deleted',	
		'enablecolumns' => array (		
			'disabled' => 'hidden',
		),
		'dynamicConfigFile' => t3lib_extMgm::extPath($_EXTKEY).'tca.php',
		'iconfile'          => t3lib_extMgm::extRelPath($_EXTKEY).'icon_tx_txbildungsangebote_angebot.png',
	),
);


t3lib_div::loadTCA('tt_content');
$TCA['tt_content']['types']['list']['subtypes_excludelist'][$_EXTKEY.'_pi1']='layout,select_key,pages';
$TCA['tt_content']['types']['list']['subtypes_addlist'][$_EXTKEY.'_pi1']='pi_flexform';


t3lib_extMgm::addPlugin(array(
	'LLL:EXT:tx_bildungsangebote/locallang_db.xml:tt_content.list_type_pi1',
	$_EXTKEY . '_pi1',
	t3lib_extMgm::extRelPath($_EXTKEY) . 'ext_icon.png'
),'list_type');


// now, add your flexform xml-file
 t3lib_extMgm::addPiFlexFormValue($_EXTKEY.'_pi1', 'FILE:EXT:'.$_EXTKEY.'/flexform_ds.xml');            // new!
	// class for displaying the category tree in BE forms.
include_once(t3lib_extMgm::extPath($_EXTKEY).'pi1/class.tx_txbildungsangebote_pi1_addFieldsToFlexForm.php');

if (TYPO3_MODE == 'BE') {
	$TBE_MODULES_EXT['xMOD_db_new_content_el']['addElClasses']['tx_txbildungsangebote_pi1_wizicon'] = t3lib_extMgm::extPath($_EXTKEY).'pi1/class.tx_txbildungsangebote_pi1_wizicon.php';
}
?>

pi1/class.tx_txbildungsangebote_pi1.php
<?php
/***************************************************************
 * Copyright notice
 *
 * (c) 2011 Florian Schuhmann <schuhmann.florian@googlemail.com>
 * All rights reserved
 *
 * This script is part of the TYPO3 project. The TYPO3 project is
 * free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * The GNU General Public License can be found at
 * http://www.gnu.org/copyleft/gpl.html.
 *
 * This script is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/
/**
 * [CLASS/FUNCTION INDEX of SCRIPT]
 *
 * Hint: use extdeveval to insert/update function index above.
 */

require_once (PATH_tslib . 'class.tslib_pibase.php');

/**
 * Plugin 'Angebote nach Monaten' for the 'tx_bildungsangebote' extension.
 *
 * @author	Florian Schuhmann <schuhmann.florian@googlemail.com>
 * @package	TYPO3
 * @subpackage	tx_txbildungsangebote
 */
class tx_txbildungsangebote_pi1 extends tslib_pibase {
	var $prefixId = 'tx_txbildungsangebote_pi1'; // Same as class name
	var $scriptRelPath = 'pi1/class.tx_txbildungsangebote_pi1.php'; // Path to this script relative to the extension dir.
	var $extKey = 'tx_bildungsangebote'; // The extension key.
	var $pi_checkCHash = true;
	
	/**
	 * Main method of your PlugIn
	 *
	 * @param	string		$content: The content of the PlugIn
	 * @param	array		$conf: The PlugIn Configuration
	 * @return	The content that should be displayed on the website
	 */
	function main($content, $conf) {
		$this->pi_initPIflexForm (); // Init and get the flexform data of the plugin
		$piFlexForm = $this->cObj->data ['pi_flexform'];
		
		$this->lConf = array ();
		foreach ( $piFlexForm ['data'] as $sheet => $data ) {
			foreach ( $data as $lang => $value ) {
				foreach ( $value as $key => $val ) {
					$this->lConf [$key] = $this->pi_getFFvalue ( $piFlexForm, $key, $sheet );
				}
			}
		}
		
		if ($this->lConf ['what_to_display'] == "MONTH") {
			$js .= '<script src="' . t3lib_extMgm::siteRelPath ( $this->extKey ) . 'static/js/SpryTabbedPanels.js" type="text/javascript"><!-- //--></script>';
			$GLOBALS ['TSFE']->additionalHeaderData [$this->extKey] = $js;
			$GLOBALS ['TSFE']->additionalHeaderData [$this->extKey . '_css_1'] = '<link href="' . t3lib_extMgm::siteRelPath ( $this->extKey ) . 'static/css/stylesheet.css" rel="stylesheet" type="text/css" />';
			
			return $this->pi_wrapInBaseClass ( $this->monthView ( $content, $conf ) );
		
		} else if ($this->lConf ['what_to_display'] == "LIST") {
			$js .= '<script src="' . t3lib_extMgm::siteRelPath ( $this->extKey ) . 'static/js/SpryCollapsiblePanel.js" type="text/javascript"><!-- //--></script>';
			$GLOBALS ['TSFE']->additionalHeaderData [$this->extKey] = $js;
			$GLOBALS ['TSFE']->additionalHeaderData [$this->extKey . '_css_1'] = '<link href="' . t3lib_extMgm::siteRelPath ( $this->extKey ) . 'static/css/SpryCollapsiblePanel.css" rel="stylesheet" type="text/css" />';
			
			return $this->pi_wrapInBaseClass ( $this->listView ( $content, $conf ) );
		} else if ($this->lConf ['what_to_display'] == "LIST2") {
			return $this->pi_wrapInBaseClass ( $this->listView2 ( $content, $conf ) );
		} else if ($this->lConf ['what_to_display'] == "LIST3") {
			return $this->pi_wrapInBaseClass ( $this->listView3 ( $content, $conf ) );
		}
	}
	
	/**
	 * Shows a month view of database entries
	 *
	 * @param	string		$content: content of the PlugIn
	 * @param	array		$conf: PlugIn Configuration
	 * @return	HTML list of table entries
	 */
	function monthView($content, $conf) {
		$this->conf = $conf; // Setting the TypoScript passed to this function in $this->conf
		$this->pi_setPiVarDefaults ();
		
		$this->pi_loadLL (); // Loading the LOCAL_LANG values
		

		$this->month = array (1 => "Jan", 2 => "Feb", 3 => "M&auml;rz", 4 => "April", 5 => "Mai", 6 => "Juni", 7 => "Juli", 8 => "Aug", 9 => "Sep", 10 => "Okt", 11 => "Nov", 12 => "Dez" );
		
		//DEBUG
		//$fullTable = var_dump ( $tRows );
		$timeShift = 10;
		$contentArray = $this->getSortedArray ( $timeShift );
		
		$navigation = '<ul class="TabbedPanelsTabGroup">';
		$tabs = '<div class="TabbedPanelsContentGroup">';
		
		for($i = date ( "n" ); $i <= date ( "n" ) + $timeShift; $i ++) {
			$tmp = date ( "n", mktime ( 0, 0, 0, $i, date ( 'y' ) ) );
            if(count($contentArray [$tmp]) > 0)
            {
			    $monat = $this->month [$tmp];
			    $navigation .= '<li tabindex="0" class="TabbedPanelsTab">' . $monat . '</li>';
			    $tabs .= $this->getTab ( $contentArray [$tmp] );
            }
		}
		$navigation .= '</ul>';
		$tabs .= '</div>';
		
		/**
		 * DEBUG: Output the content of $this->piVars for debug purposes. 
		 * REMEMBER to comment out the IP-lock in the debug() function in t3lib/config_default.php 
		 * if nothing happens when you un-comment this line! 
		 */
		//$fullTable.=t3lib_div::view_array($this->lConf);	
		

		// Returns the content from the plugin.
		$js = '<script type="text/javascript">var TabbedPanels1 = new Spry.Widget.TabbedPanels("Termine");</script>';
		return '<div id="Termine" class="TabbedPanels">' . $navigation . $tabs . '</div>' . $js;
	
	}
	/**
	 * Shows a list view of database entries
	 *
	 * @param	string		$content: content of the PlugIn
	 * @param	array		$conf: PlugIn Configuration
	 * @return	HTML list of table entries
	 */
	function listView($content, $conf) {
		$this->conf = $conf; // Setting the TypoScript passed to this function in $this->conf
		$this->pi_setPiVarDefaults ();
		
		$this->pi_loadLL (); // Loading the LOCAL_LANG values
		

		//DEBUG
		//$fullTable = var_dump ( $tabs );		
		

		$contentArray = $this->getListArray ();
		$tabs = $this->getItems ( $contentArray );
		
		$uid = $this->cObj->data ['uid'];
		$text = '<span class="textbg">' . $this->lConf['subText'] . '</span>';
		$js = '<script type="text/javascript">var CollapsiblePanel' . $uid . ' = new Spry.Widget.CollapsiblePanel("CollapsiblePanel' . $uid . '", {contentIsOpen:true});</script>';
		return '<div id="CollapsiblePanel' . $uid . '" class="CollapsiblePanel">' . $tabs . '</div>' . $text . $js;
	
	}
	/**
	 * Shows a list view2 of database entries
	 *
	 * @param	string		$content: content of the PlugIn
	 * @param	array		$conf: PlugIn Configuration
	 * @return	HTML list of table entries
	 */
	function listView2($content, $conf) {
		$this->conf = $conf; // Setting the TypoScript passed to this function in $this->conf
		$this->pi_setPiVarDefaults ();
		
		$this->pi_loadLL (); // Loading the LOCAL_LANG values
		

		//DEBUG
		//$fullTable = var_dump ( $tabs );		
		$contentArray = $this->getListArray ();
		$temp = $this->getItemsWithImages ( $contentArray );
		
		return '<ul class="ba">' . $temp . '</ul>';
	
	}	
	/**
	 * Shows a list view3 of database entries
	 *
	 * @param	string		$content: content of the PlugIn
	 * @param	array		$conf: PlugIn Configuration
	 * @return	HTML list of table entries
	 */
	function listView3($content, $conf) {
		$this->conf = $conf; // Setting the TypoScript passed to this function in $this->conf
		$this->pi_setPiVarDefaults ();
		
		$this->pi_loadLL (); // Loading the LOCAL_LANG values
		

		//DEBUG
		//$fullTable = var_dump ( $tabs );		
		$contentArray = $this->getListArray ();
		return $temp = $this->getItemsBoxed ( $contentArray );
	}	
	function getStartDateFormated($start) {
		return strftime ( '%d.%m.%Y', $start );
	
	}
	function getTab($month) {
		
		$tab = '<div class="TabbedPanelsContent">';
		
		foreach ( $month as $sub){
		foreach ($sub as $key => $cat ) {
		
		
			$cat_old = null;
			foreach ( $cat as $angebot ) {
			if ( $angebot ['cattitle'] == 'Mit täglichem Einstieg' or $angebot ['cattitle'] == 'Berufsbegleitend' or $angebot ['cattitle'] == 'Umschulung' or $angebot ['cattitle'] == 'Modulare Qualifizierung' or $angebot ['cattitle'] == 'Weiterbildungskurse'){
				if ($key != $cat_old)
					$tab .= '<span class="headline">' . $angebot ['cattitle'] . '</span>';
				
				$tab .= '<p>';
				
				if ($angebot ['pdf'] != "")
					$tab .= '<a target="_blank" href="' . $angebot ['pdf'] . '">';
				
				$tab .= '<strong>' . $angebot ['title'] . '</strong>';
				
				if ($angebot ['pdf'] != "")
					$tab .= '</a>';
				
				if ($angebot ['number'] != "")
					$tab .= '<br>' . $angebot ['number'];
				
				if ($angebot ['duration'] != "")
					$tab .= '<br>Dauer: ' . $angebot ['duration'];
				
				if ($angebot ['events'] != ""){
				    $parts= explode("\n", $angebot ['events']);
					$tab .= '<br><span class="date">Termin:<br>' . $parts [0] . '</span>';
				}
				
				$tab .= '</p>';
				$cat_old = $key;
			}
		}
		
		}
		}
		// DEBUG
		//$tab.=t3lib_div::view_array($month);
		$tab .= '</div>';
		return $tab;

	}
	function getItems($items) {
		$title = '<div class="CollapsiblePanelTab" tabindex="0"><span class="headlinebg">';
		$content = '<div class="CollapsiblePanelContent">';
		if ( $angebot ['cattitle'] == 'Mit täglichem Einstieg' or $angebot ['cattitle'] == 'Berufsbegleitend' or $angebot ['cattitle'] == 'Umschulung' or $angebot ['cattitle'] == 'Modulare Qualifizierungen' or $angebot ['cattitle'] == 'Weiterbildungskurse' ){
		foreach ( $items as $key => $angebot ) {
			if ($key == 0)
				$title .= $angebot ['cattitle'];
			
			
			
			$content .= '<div class="angebot">';
			
			if ($angebot ['pdf'] != "")
				$content .= '<a target="_blank" href="' . $angebot ['pdf'] . '">';
			
			$content .= '<strong>' . $angebot ['title'] . '</strong>';
			
			if ($angebot ['pdf'] != "")
				$content .= '</a>';
			
			
			if ($angebot ['number'] != "")
				$content .= '<br>' . $angebot ['number'];
			
			if ($angebot ['duration'] != "")
				$content .= '<br>Dauer: ' . $angebot ['duration'];
			
			if ($angebot ['events'] != "")
				$content .= '<br><span class="date">Termin:<br>' . nl2br($angebot ['events']) . '</span>';
			
			$content .= '</div>';
		}
		// DEBUG
		// $content .= t3lib_div::view_array ( $items );
		$title .= '</span></div>';
		$content .= '</div>';
		return $title . $content;
	}
	}
	function getItemsWithImages($items) {
		foreach ( $items as $key => $angebot ) {
		$content .= '<li>';
		    // IMAGE
			if ($angebot ['pdf'] != "" &&  $angebot ['image'] != "")
				$content .= '<div class="left"><a target="_blank" href="' . $angebot ['pdf'] . '">';
			
			if ($angebot ['image'] != ""){
			    $path=(!preg_match("/fileadmin/", $angebot ['image']))?'uploads/tx_txbildungsangebote/':'';
			    $content .= '<img class="border" height="90" width="90" src="' . $path . $angebot ['image'] . '"/>';
			}
			
			if ($angebot ['pdf'] != "" &&  $angebot ['image'] != "")
				$content .= '</a></div>';
		    
			if ($angebot ['pdf'] != "")
				$content .= '<div class="right"><a target="_blank" href="' . $angebot ['pdf'] . '">';
			
			$content .= '<strong>' . $angebot ['title'] . '</strong>';
			
			if ($angebot ['pdf'] != "")
				$content .= '</a>';
			
			if ($angebot ['description'] != "")
				$content .= '<br>' . $angebot ['description'].'<br>';

			if ($angebot ['events'] != "")
				$content .= '<br><span class="nornal">Termin:<br>' . nl2br($angebot ['events']) . '</span>';
				
				if ($angebot ['duration'] != "")
				$content .= 'Dauer: ' . $angebot ['duration'];
				
		$content .= '</div></li>';			
		}
		// DEBUG
		// $content .= t3lib_div::view_array ( $items );
		return $content;
	
	}
	function getItemsBoxed($items) {
	    $content ='<div class="module_container">';
		foreach ( $items as $key => $angebot ) {
			if ($key == 0)
				$title = $angebot ['cattitle'];
		    
			$content .= '<div class="module_box">';

			if($angebot['color'] == '') $angebot['color'] = '#0E555B';
			$content .= '<div class="top" style="color:'.$angebot['color'].'">' .
			            $angebot ['title'] . '</div>';

			$content .='<div class="top_content" >';
			if ($angebot ['description'] != "")
				$content .= $angebot ['description'];
			
			if ($angebot ['events'] != "")
				$content .= '<br><br><span class="date">Termin:<br>' . nl2br($angebot ['events']) . '</span>';
			
			if ($angebot ['duration'] != "")
				$content .= '<br><span class="nornal">Dauer: ' . $angebot ['duration'] . '</span>';
			
			$content .= '</div>';
			
			$content .= '<div class="module_box_link" >';
			if ($angebot ['pdf'] != "")
				$content .= '<a target="_blank" href="' . $angebot ['pdf'] . '">Download Flyer &raquo;</a>';
            $content .= '</div>';


			
		$content .= '</div>';			
		}
		$content .= '</div>';
		// DEBUG
		// $content .= t3lib_div::view_array ( $items );
		return $content;
	
	}
	function getSortedArray($timeShift) {
		
		$from_date = mktime ( 0, 0, 0, date ( 'm' ), 1, date ( 'y' ) ); // todays date
		

		$tmp = mktime ( 0, 0, 0, date ( 'm' ) + $timeShift, 1, date ( 'y' ) );
		$to_date = mktime ( 0, 0, 0, date ( 'm' ) + $timeShift, date ( 't', $tmp ), date ( 'y' ) ); // date with one year difference i.e. same date of next year
		

		//$GLOBALS['TYPO3_DB']->debugOutput = TRUE; 
		$res = $GLOBALS ['TYPO3_DB']->exec_SELECTquery ( 'DISTINCT c.sorting as sort, a.*, c.uid as cat, FROM_UNIXTIME(a.start,"%m") as monat, c.title as cattitle', // SELECT ...
'tx_txbildungsangebote_angebot a, tx_txbildungsangebote_cat c, tx_txbildungsangebote_angebot_category_mm mm ', // FROM ...
'a.start >= "' . $from_date . '" and a.start <= "' . $to_date . '"' . ' and dayly_repeat=0 and mm.uid_foreign=c.uid' . ' and mm.uid_local=a.uid' . ' and not a.hidden and not a.deleted' . // all offers that are NOT deleted and NOT hidden
' and not c.hidden and not c.deleted' . ' and a.pid="' . $this->lConf ['pages'] . '"', // all categories that are NOT deleted and NOT hidden
'', // GROUP BY... 
'a.sorting asc, c.sorting asc, monat asc, cat asc', // ORDER BY...
'' ); // LIMIT to 10 rows, starting with number 5 (MySQL compat.)
		

		$tRows = null;
		$c = array ();
		while ( $row = $GLOBALS ['TYPO3_DB']->sql_fetch_assoc ( $res ) ) {
			$c2 = $c [$row ['monat']];
			$c2 = ($c2) ? $c2 : 0;
			echo($row['sort']);
			$tRows [intval ( $row ['monat'] )][$row['sort']] [$row ['cat']] [$c2] = $row;
			$c [$row ['monat']] ++;
		}
		
		
			$res = $GLOBALS ['TYPO3_DB']->exec_SELECTquery ( 'DISTINCT  c.sorting as sort,a.*, c.uid as cat, c.title as cattitle', // SELECT ...
'tx_txbildungsangebote_angebot a, tx_txbildungsangebote_cat c, tx_txbildungsangebote_angebot_category_mm mm ', // FROM ...
'a.dayly_repeat=1' . ' and mm.uid_foreign=c.uid' . ' and mm.uid_local=a.uid' . ' and not a.hidden and not a.deleted' . // all offers that are NOT deleted and NOT hidden
' and not c.hidden and not c.deleted' . ' and a.pid="' . $this->lConf ['pages'] . '"', // all categories that are NOT deleted and NOT hidden
'', // GROUP BY... 
'a.sorting asc, c.sorting asc, cat asc', // ORDER BY...
'' ); // LIMIT to 10 rows, starting with number 5 (MySQL compat.)
		

		while ( $row = $GLOBALS ['TYPO3_DB']->sql_fetch_assoc ( $res ) ) {
		    $from_date = mktime ( 0, 0, 0, date ( 'm' ), 1, date ( 'y' ) ); // todays date
		    $i=1;
            while ($from_date <$to_date){
                $tRows[date('n',$from_date)][$row['sort']][$row['cat']][] = $row;
                ksort($tRows[date('n',$from_date)]);
                
                $from_date = mktime ( 0, 0, 0, date ( 'm' )+$i, 1, date ( 'y' ) ); // todays date
                $i++;
            }
            
		}
		
		//die(t3lib_div::view_array ($tRows ));
		return $tRows;
	}
	
	function getListArray() {
		
		$from_date = mktime ( 0, 0, 0, date ( 'm' ), 1, date ( 'y' ) ); // todays date
		

		//$GLOBALS['TYPO3_DB']->debugOutput = TRUE; 
		$res = $GLOBALS ['TYPO3_DB']->exec_SELECTquery ( 'DISTINCT a.*, c.uid as cat, FROM_UNIXTIME(a.start,"%m") as monat, c.title as cattitle', // SELECT ...
'tx_txbildungsangebote_angebot a, tx_txbildungsangebote_cat c, tx_txbildungsangebote_angebot_category_mm mm ', // FROM ...
'(a.start >= "' . $from_date . '" or a.start = "0")' . ' and mm.uid_foreign=c.uid' . ' and mm.uid_local=a.uid' . ' and not a.hidden and not a.deleted' . // all offers that are NOT deleted and NOT hidden
' and not c.hidden and not c.deleted' . // all categories that are NOT deleted and NOT hidden 
' and a.pid="' . $this->lConf ['pages'] . '"' . ' and c.uid="' . $this->lConf ['categorySelection'] . '"', '', // GROUP BY... 
'a.sorting asc, c.sorting asc, monat asc, cat asc', // ORDER BY...
'' ); // LIMIT to 10 rows, starting with number 5 (MySQL compat.)
		

		$tRows = null;
		while ( $row = $GLOBALS ['TYPO3_DB']->sql_fetch_assoc ( $res ) ) {
			$tRows [] = $row;
		}
		return $tRows;
	}
}

if (defined ( 'TYPO3_MODE' ) && $TYPO3_CONF_VARS [TYPO3_MODE] ['XCLASS'] ['ext/tx_bildungsangebote/pi1/class.tx_txbildungsangebote_pi1.php']) {
	include_once ($TYPO3_CONF_VARS [TYPO3_MODE] ['XCLASS'] ['ext/tx_bildungsangebote/pi1/class.tx_txbildungsangebote_pi1.php']);
}

?>

tca.php
<?php
if (!defined('TYPO3_MODE')) {
	die ('Access denied.');
}

$TCA['tx_txbildungsangebote_cat'] = array (
	'ctrl' => $TCA['tx_txbildungsangebote_cat']['ctrl'],
	'interface' => array (
		'showRecordFieldList' => 'hidden,title,parent_category'
	),
	'feInterface' => $TCA['tx_txbildungsangebote_cat']['feInterface'],
	'columns' => array (
		'hidden' => array (		
			'exclude' => 1,
			'label'   => 'LLL:EXT:lang/locallang_general.xml:LGL.hidden',
			'config'  => array (
				'type'    => 'check',
				'default' => '0'
			)
		),
		'title' => array (		
			'exclude' => 0,		
			'label' => 'LLL:EXT:tx_bildungsangebote/locallang_db.xml:tx_txbildungsangebote_cat.title',		
			'config' => array (
				'type' => 'input',	
				'size' => '30',	
				'eval' => 'required',
			)
		),
		'parent_category' => array (		
			'exclude' => 0,		
			'label' => 'LLL:EXT:tx_bildungsangebote/locallang_db.xml:tx_txbildungsangebote_cat.parent_category',		
			'config' => array (
				'type' => 'group',	
				'internal_type' => 'db',	
				'allowed' => 'tx_txbildungsangebote_cat',	
				'size' => 1,	
				'minitems' => 0,
				'maxitems' => 1,
			)
		),
	),
	'types' => array (
		'0' => array('showitem' => 'hidden;;1;;1-1-1, title;;;;2-2-2, parent_category;;;;3-3-3')
	),
	'palettes' => array (
		'1' => array('showitem' => '')
	)
);



$TCA['tx_txbildungsangebote_angebot'] = array (
	'ctrl' => $TCA['tx_txbildungsangebote_angebot']['ctrl'],
	'interface' => array (
		'showRecordFieldList' => 'hidden,title,description,duration,start,category,pdf,number,dayly_repeat,image,events,color'
	),
	'feInterface' => $TCA['tx_txbildungsangebote_angebot']['feInterface'],
	'columns' => array (
		'hidden' => array (		
			'exclude' => 1,
			'label'   => 'LLL:EXT:lang/locallang_general.xml:LGL.hidden',
			'config'  => array (
				'type'    => 'check',
				'default' => '0'
			)
		),
		'title' => array (		
			'exclude' => 0,		
			'label' => 'LLL:EXT:tx_bildungsangebote/locallang_db.xml:tx_txbildungsangebote_angebot.title',		
			'config' => array (
				'type' => 'input',	
				'size' => '30',	
				'eval' => 'required',
			)
		),
		'description' => array (		
			'exclude' => 0,		
			'label' => 'LLL:EXT:tx_bildungsangebote/locallang_db.xml:tx_txbildungsangebote_angebot.description',		
			'config' => array (
				'type' => 'text',
				'cols' => '30',
				'rows' => '5',
				'wizards' => array(
					'_PADDING' => 2,
					'RTE' => array(
						'notNewRecords' => 1,
						'RTEonly'       => 1,
						'type'          => 'script',
						'title'         => 'Full screen Rich Text Editing|Formatteret redigering i hele vinduet',
						'icon'          => 'wizard_rte2.gif',
						'script'        => 'wizard_rte.php',
					),
				),
			)
		),
		'duration' => array (		
			'exclude' => 0,		
			'label' => 'LLL:EXT:tx_bildungsangebote/locallang_db.xml:tx_txbildungsangebote_angebot.duration',		
			'config' => array (
				'type' => 'input',	
				'size' => '30',
			)
		),
		'start' => array (		
			'exclude' => 0,		
			'label' => 'LLL:EXT:tx_bildungsangebote/locallang_db.xml:tx_txbildungsangebote_angebot.start',		
			'config' => array (
				'type'     => 'input',
				'size'     => '8',
				'max'      => '20',
				'eval'     => 'date',
				'checkbox' => '0',
				'default'  => '0'
			)
		),
		'category' => array (		
			'exclude' => 0,		
			'label' => 'LLL:EXT:tx_bildungsangebote/locallang_db.xml:tx_txbildungsangebote_angebot.category',		
			'config' => array (
				'type' => 'group',	
				'internal_type' => 'db',	
				'allowed' => 'tx_txbildungsangebote_cat',	
				'size' => 3,	
				'minitems' => 0,
				'maxitems' => 5,	
				"MM" => "tx_txbildungsangebote_angebot_category_mm",
			)
		),
		'pdf' => array (		
			'exclude' => 0,		
			'label' => 'LLL:EXT:tx_bildungsangebote/locallang_db.xml:tx_txbildungsangebote_angebot.pdf',		
			'config' => array (
				'type'     => 'input',
				'size'     => '15',
				'max'      => '255',
				'checkbox' => '',
				'eval'     => 'trim',
				'wizards'  => array(
					'_PADDING' => 2,
					'link'     => array(
						'type'         => 'popup',
						'title'        => 'Link',
						'icon'         => 'link_popup.gif',
						'script'       => 'browse_links.php?mode=wizard',
						'JSopenParams' => 'height=300,width=500,status=0,menubar=0,scrollbars=1'
					)
				)
			)
		),
		'number' => array (		
			'exclude' => 0,		
			'label' => 'LLL:EXT:tx_bildungsangebote/locallang_db.xml:tx_txbildungsangebote_angebot.number',		
			'config' => array (
				'type' => 'input',	
				'size' => '30',
			)
		),
		'dayly_repeat' => array (		
			'exclude' => 0,		
			'label' => 'LLL:EXT:tx_bildungsangebote/locallang_db.xml:tx_txbildungsangebote_angebot.dayly_repeat',		
			'config' => array (
				'type' => 'check',
			)
		),
		'image' => array (		
			'exclude' => 0,		
			'label' => 'LLL:EXT:tx_bildungsangebote/locallang_db.xml:tx_txbildungsangebote_angebot.image',		
			'config' => array (
				'type' => 'group',
				'internal_type' => 'file',
				'allowed' => 'gif,png,jpeg,jpg',	
				'max_size' => $GLOBALS['TYPO3_CONF_VARS']['BE']['maxFileSize'],	
				'uploadfolder' => 'uploads/tx_txbildungsangebote',
				'show_thumbs' => 1,	
				'size' => 1,	
				'minitems' => 0,
				'maxitems' => 1,
			)
		),
		'events' => array (		
			'exclude' => 0,		
			'label' => 'LLL:EXT:tx_bildungsangebote/locallang_db.xml:tx_txbildungsangebote_angebot.events',		
			'config' => array (
				'type' => 'text',
				'cols' => '30',	
				'rows' => '5',
			)
		),
		'color' => array (		
			'exclude' => 0,		
			'label' => 'LLL:EXT:tx_bildungsangebote/locallang_db.xml:tx_txbildungsangebote_angebot.color',		
			'config' => array (
				'type' => 'input',	
				'size' => '10',	
				'max' => '7',	
				'wizards' => array(
					'_PADDING' => 2,
					'color' => array(
						'title' => 'Color:',
						'type' => 'colorbox',
						'dim' => '12x12',
						'tableStyle' => 'border:solid 1px black;',
						'script' => 'wizard_colorpicker.php',
						'JSopenParams' => 'height=300,width=250,status=0,menubar=0,scrollbars=1',
					),
				),
				'eval' => 'nospace',
			)
		),
	),
	'types' => array (
		'0' => array('showitem' => 'hidden;;1;;1-1-1, title;;;;2-2-2, description;;;richtext[]:rte_transform[mode=ts];3-3-3, duration, start, category, pdf, number, dayly_repeat, image, events, color')
	),
	'palettes' => array (
		'1' => array('showitem' => '')
	)
);
?>